contracts/FlowContractAudits.cdc

```diff
@@ -0,0 +1,215 @@
+    }
+
+    // The auditor account will have audit access through AuditorProxy
+    // This enables the admin account to revoke access
+    // See https://docs.onflow.org/cadence/design-patterns/#capability-revocation
+    pub resource AuditorProxy: AuditorProxyPublic {
+        access(self) var auditorCapability: Capability<&Auditor>?
+
+        pub fun setAuditorCapability(_ cap: Capability<&Auditor>) {
+            self.auditorCapability = cap
+        }
+
+        pub fun addVoucher(address: Address?, recurrent: Bool, expiryOffset: UInt64?, code: String) {
+            self.auditorCapability!.borrow()!.addVoucher(address: address, recurrent: recurrent, expiryOffset: expiryOffset, code: code)
+        }
+
+        pub fun deleteVoucher(key: String) {
+            self.auditorCapability!.borrow()!.deleteVoucher(key: key)
+        }
+
+        init() {
+            self.auditorCapability = nil
+        }
+
+    }
+
+    // Can be called by anyone but needs a capability to function
+    pub fun createAuditorProxy(): @AuditorProxy {
+        return <- create AuditorProxy()
+    }
+
+    pub resource Administrator {
+        
+        // Creates new Auditor
+        pub fun createNewAuditor(): @Auditor {
+            emit AuditorCreated()
+            return <-create Auditor()
+        }
+
+        // Checks all vouchers and removes expired ones
+        pub fun cleanupExpiredVouchers() {
+            for key in FlowContractAudits.vouchers.keys {                
+                let v = FlowContractAudits.vouchers[key]!
+                if v.expiryBlockHeight != nil {
+                    if getCurrentBlock().height > v.expiryBlockHeight! {
+                        FlowContractAudits.vouchers.remove(key: key)
+                        emit VoucherRemoved(key: key, recurrent: v.recurrent, expiryBlockHeight: v.expiryBlockHeight)                
+                    }
+                }
+            }
+        }
+
+        // For testing
+        pub fun useVoucherForDeploy(address: Address, code: String): Bool {
+            return FlowContractAudits.useVoucherForDeploy(address: address, code: code)
+        }
+    }
+
+    // This function will be called by the FVM on contract deploy/update
+    access(contract) fun useVoucherForDeploy(address: Address, code: String): Bool {
+        let codeHash = FlowContractAudits.hashContractCode(code)
+        var key = FlowContractAudits.generateVoucherKey(address: address, codeHash: codeHash)
+
+        // first check for voucher based on target account
+        // if not found check for any account
+        if !FlowContractAudits.vouchers.containsKey(key) {
+            key = FlowContractAudits.generateVoucherKey(address: nil, codeHash: codeHash)
+            if !FlowContractAudits.vouchers.containsKey(key) {
+                return false
+            }
+        }
+
+        let v = FlowContractAudits.vouchers[key]!
+
+        // ensure contract code matches the voucher
+        if v.codeHash == codeHash  {
```

Comment:
It might be better to invert the if statement to reduce nesting.

Reply:
Good idea! Addressed here https://github.com/onflow/flow-core-contracts/pull/271/commits/8d7a1e5bda44a3372b4fb2ae1542b2e6a99961ed